File: 백준/Gold/1025. 제곱수 찾기/제곱수 찾기.py
import math
import sys

NM = sys.stdin.readline().split()
N = int(NM[0])
M = int(NM[1])
ans = -1

graph = [[0 for _ in range(M)]for _ in range(N)]
for n in range(N):
	ent = sys.stdin.readline()
	for m in range(M):
		graph[n][m] = int(ent[m])

def slv(r, c):
	for i in range(-N,N):
		for j in range(-M,M):
			if(i==0 and j==0):
				continue

			global ans
			x=r
			y=c
			sqr=0

			while 0<=x<N and 0<=y<M:
				sqr *= 10
				sqr += graph[x][y]
				if math.sqrt(sqr)-int(math.sqrt(sqr))==0:
					ans = ans if ans>sqr else sqr

				x+=i
				y+=j

for n in range(N):
	for m in range(M):
		slv(n,m)

print(ans)

File: 백준/Gold/1025. 제곱수 찾기/제곱수 찾기.java
import java.io.*;
import java.util.*;

public class Main {
    static int N, M;
    static int[][] map;
    static int ans = -1;

    public static void main(String[] args) throws IOException {
        //System.setIn(new FileInputStream("src/input.txt"));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        map = new int[N][M];
        for(int n=0; n<N; n++){
            String str = br.readLine();
            for(int m=0; m<M; m++){
                map[n][m] = str.charAt(m) - '0';
            }
        }

        for(int i=0; i<N; ++i){
            for(int j=0; j<M; ++j){
                slv(j, i);
            }
        }
        System.out.println(ans);
    }

    public static void slv(int c, int r){
        for(int i=-N; i<N; ++i){
            for(int j=-M; j<M; ++j){
                if(i == 0 && j == 0) continue;

                int x = c;
                int y = r; 

                int sqr = 0;

                while (0<=x && x<M && 0<=y && y<N){
                    sqr *= 10;
                    sqr += map[y][x];

                    if( Math.sqrt(sqr)-(int)Math.sqrt(sqr) == 0 )
                        ans = Math.max(ans, sqr);

                    x += j;
                    y += i;
                }
            }
        }
    }
}